src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                newOptions = additionalOptions;
+            }
+            if (
+                selectedOption &&
+                !additionalOptions?.find((o) => o.text === selectedOption.text)
+            ) {
+                newOptions = newOptions.concat([selectedOption]);
+            }
+        }
+
+        setOptions(newOptions);
+    }, [resourcesState.adapterResult]);
+
+    useEffect(() => {
+        if (selectedOptionProp) {
+            setSelectedOption({
+                key: selectedOptionProp,
+                text: selectedOptionProp,
+                styles: comboBoxOptionStyles
+            });
+        } else {
+            setSelectedOption(null);
+        }
+    }, [selectedOptionProp]);
+
+    useEffect(() => {
+        if (selectedOption) {
+            const selectedKey = options?.find(
+                (option) =>
+                    option.text === parsedOptionText(selectedOption.text)
+            )?.key;
+            setSelectedKey(selectedKey);
+        } else {
+            setSelectedKey(null);
+        }
+    }, [selectedOption]);
+
+    useEffect(() => {
+        setAdditionalOptions(
+            additionalOptionsProp?.map((aO) => ({
+                key: aO,
+                text: aO,
+                styles: comboBoxOptionStyles
+            })) || []
+        );
+    }, [additionalOptionsProp]);
+
+    useEffect(() => {
+        setOptions(
+            mergeOptionsWithProps(
+                optionsFromResources(),
+                additionalOptions,
+                selectedOption
+            )
+        );
+    }, [additionalOptions]);
+
+    const parsedOptionText = useCallback((value) => {
+        if (!value) return value;
+
+        if (displayField === AzureResourceDisplayFields.url) {
+            // let user enter hostname and gracefully append https protocol
+            let newVal = value;
+            if (!newVal.startsWith('https://')) {
+                if (
+                    (resourceType === AzureResourceTypes.DigitalTwinInstance &&
+                        ValidAdtHostSuffixes.some(
+                            (suffix) =>
+                                newVal.endsWith(suffix) ||
+                                newVal.endsWith(suffix + '/')
+                        )) ||
+                    ((resourceType ===
+                        AzureResourceTypes.StorageBlobContainer ||
+                        resourceType === AzureResourceTypes.StorageAccount) &&
+                        ValidContainerHostSuffixes.some(
+                            (suffix) =>
+                                newVal.endsWith(suffix) ||
+                                newVal.endsWith(suffix + '/')
+                        ))
+                ) {
+                    newVal = 'https://' + newVal;
+                }
+            }
+            return newVal;
+        } else {
+            return value;
+        }
+    }, []);
+
+    const isValidUrlStr = useCallback((urlStr: string) => {
```

Comment:
let's pull this out to a helper function. It seems self contained enough and will simplify the render loop.